### PR TITLE
Timezone Fix

### LINKED ISSUE #1745: Submission timezone not honored
Our server is set to PST and I set the submission deadline to midnight PST (GMT - 8). Yet the submission time shows as three hours later and the students are told that their submissions are late if they are after 9pm.

### PR DESCRIPTION
**Pull request by RylanOC**
Closes #1745 

The timezone is no longer changed in `install_system.sh`, and instead only ever modified in `Vagrantfile`.

PHP doesn't automatically inherit the timezone from the server, so any scripts that deal with time sensitive calculates needed to be updated to read from `/etc/timezone`.

Finally, `config.php` (`site/app/models/Config.php`) was hardcoded as 'America/New_York", and needed to be updated to keep the site from crashing when other timezones are specified in different PHP scripts.

### CODE REVIEW

**Comment by MasterOdin** [comment]
This will put a hard requirement of that you HAVE to use Ubuntu/Debian for Submitty, else this will not work and I'm not convinced that's where we want to be going.

Additionally, `site/app/models/Config.php` was not hardcoded to always be `America/New_York`, that's just what the default value was (as it's where we are), but it's overwritten here:
https://github.com/Submitty/Submitty/blob/c859164f49e6d49b482ded534fec843836d15730/site/app/models/Config.php#L223-L228

which will (currently) always catch as the `site/config/master_template.ini` file has this:
https://github.com/Submitty/Submitty/blob/8788391ff982704184181c4bb7c1d64073c9ac50/site/config/master_template.ini#L9

If anything, I'd prefer that we add a question within `.setup/CONFIGURE_SUBMITTY.py` that asks what timezone you want and then writes that to the to the `master.ini` file which PHP then reads and then parses all incoming datetimes into that specific timezone.

File: site/app/views/NavigationView.php [hunk 1]
```diff
@@ -5,6 +5,14 @@
...
 use app\models\Gradeable;
 
+$default_zone = "America/New_York";
+$timezone = file_get_contents("/etc/timezone");
+if (!strlen($timezone)) {
+    $timezone = $default_zone;
+}
+$timezone = trim($timezone);
+date_default_timezone_set($timezone);
+
```
Comment (MasterOdin) [inline site/app/views/NavigationView.php]: Unnecessary as we're already setting the `date_default_timezone_set` within `site/public/index.php`.

You also shouldn't introduce 'side-effects' when loading a file that has a class in it per [PSR-1](https://www.php-fig.org/psr/psr-1/#23-side-effects).

**Comment by MasterOdin** [comment]
So I believe the consensus we had was that within CONFIGURE_Submitty.py we write out the timezone to a config file and that we then reference that config file anywhere we're writing or displaying a datetime. That way if someone reconfigures the system (or re-runs install_system.sh), that'll pick up the changes to timezone for them.

@bmcutler, @RylanOC, if that's what we're good on doing, I'll add the necessary field to a file inside the conf file inside of the #1756 PR, as it should just be accomplished by doing:
```
import tzlocal
tzlocal.get_localzone().zone
```
which will give us a string that can be used within PHP or Python.

File: Docs/student_auto_feed/accounts.php [hunk 2]
```diff
@@ -53,17 +53,9 @@
  * -------------------------------------------------------------------------- */
 
 //Univeristy campus's timezone.
-date_default_timezone_set('America/New_York');
+$core = new Core();
+date_default_timezone_set($core->getConfig()->getTimezone()->getName());
```
[the remark this answers is not in the document]
Comment (pbailie) [reply Docs/student_auto_feed/accounts.php]: _At minimum_, if we don't set the timezone, PHP throws a lengthy warning about reliability.

File: Docs/student_auto_feed/config.php [hunk 3]
```diff
@@ -36,7 +38,9 @@
  * -------------------------------------------------------------------------- */
 
 // Univeristy campus's timezone.  ***THIS NEEDS TO BE SET.
-date_default_timezone_set('America/New_York');
+$core = new Core();
+date_default_timezone_set($core->getConfig()->getTimezone()->getName());
```
[the remark this answers is not in the document]
Comment (pbailie) [reply Docs/student_auto_feed/config.php]: `config.php` runs on a separate server maintained by labstaff and supports the nightly auto feed process.  There is a subtle flaw in PHP, as specifically compiled by Canonical, that makes remote file access to the feed CSV undesirable.

`accounts.php` is run on the Submitty server, but it resides with other sysadmin tools in a place that requires root access -- so not in the same place as the Submitty site code.

[the remark this answers is not in the document]
Comment (MasterOdin) [reply Docs/student_auto_feed/config.php]: Based off this, I'd just leave it setting `date_default_timezone_set('America/New_York');` as setting that whole thing up is documented elsewhere and not try and do anything with `Core` or Submitty at large here.

File: site/app/models/Config.php [hunk 4]
```diff
@@ -95,7 +95,7 @@ class Config extends AbstractModel {
     /** @property @var string */
     protected $authentication;
     /** @property @var string */
-    protected $timezone = "America/New_York";
+    protected $timezone = 'America/New_York'; //Default to EST
```
Comment (MasterOdin) [inline site/app/models/Config.php]: Remove comment as it's not really telling anything you can't just glean from looking at the default value. It's also wrong half the year when 'America/New_York' stands for EDT timezone (like right now).

File: site/app/views/NavigationView.php [hunk 5]
```diff
@@ -1,10 +1,10 @@
 <?php
-
 namespace app\views;
-
 use \app\libraries\GradeableType;
+use app\models\Config;
 use app\models\Gradeable;
-
+use app\libraries\Core;
+//date_default_timezone_set($this->core->getConfig()->getTimezone());
```
Comment (MasterOdin) [inline site/app/views/NavigationView.php]: remove this commented out line. We'll never need to do this here. 

We can also probably get rid of the
```
use app\models\Config;
use app\libraries\Core;
```
lines.

File: site/public/index.php [hunk 6]
```diff
@@ -106,7 +106,9 @@ function error_handler() {
     $core->getOutput()->addBreadcrumb("", $core->getConfig()->getCourseHomeUrl(),false, true);
 }
 
+echo $core->getConfig()->getTimezone()->getName();
```
Comment (MasterOdin) [inline site/public/index.php]: Remove this after done demoing this.

**Comment by RylanOC** [comment]
@MasterOdin @pbailie 
Thanks for your feedback, I think I've made all the fixes that you requested. If you spot any other issues I'd love to know.

File: Docs/student_auto_feed/config.php [hunk 7]
```diff
@@ -1,5 +1,7 @@
 <?php
 
+use app\libraries\Core;
```
Comment (MasterOdin) [inline Docs/student_auto_feed/config.php]: Remove this

File: .setup/CONFIGURE_SUBMITTY.py [hunk 8]
```diff
@@ -133,7 +131,8 @@ def get_input(question, default=""):
             'authentication_method': 1,
             'institution_name' : '',
             'username_change_text' : 'Submitty welcomes individuals of all ages, backgrounds, citizenships, disabilities, sex, education, ethnicities, family statuses, genders, gender identities, geographical locations, languages, military experience, political views, races, religions, sexual orientations, socioeconomic statuses, and work experiences. In an effort to create an inclusive environment, you may specify a preferred name to be used instead of what was provided on the registration roster.',
-            'institution_homepage' : ''}
+            'institution_homepage' : '',
+            'timezone' : tzlocal.get_localzone()}
```
Comment (MasterOdin) [inline .setup/CONFIGURE_SUBMITTY.py]: you probably want to do `tzlocal.get_localzone().zone()` here as `tzlocal.get_localzone()` returns a TzInfo object. Example:
```
>>> import tzlocal
>>> tzlocal.get_localzone()
<DstTzInfo 'America/New_York' LMT-1 day, 19:04:00 STD>
>>> tzlocal.get_localzone().zone
'America/New_York'
```

**Comment by MasterOdin** [comment]
@RylanOC I fixed a typo in that function call that I gave you (zone is a property, not a method, whoops!).
@bmcutler this looks good to merge in my opinion

**Pull request merged**